### ISSUE TITLE
ABCMeta.register support

### ISSUE THREAD

**Issue by miedzinski** [python/mypy#2922]
```python
class A(metaclass=abc.ABCMeta): pass
class B: pass
A.register(B)
a: A = B()  # currently E: Incompatible types in assignment (expression has type "B", variable has type "A")
print(isinstance(a, A))  # -> True
```

Mypy could support cases like this by special casing `ABCMeta.register` calls. `__subclasshook__`, `__subclasscheck__`, and `__instancecheck__` are more troublesome and could be left ignored.

**Comment by JukkaL** [comment]
This was on my radar early in mypy development. I didn't work on this back then because this is a tricky feature. For example, if module `x` has `A.register(B)`, it can affect a seemingly unrelated module `y` if it tries to use `B` as a subclass of `A`. This is 'non-modular' behavior, and mypy would somehow have to figure out that `y` depends on `x` even if it doesn't import it, even indirectly. Currently dependencies between modules are completely described by the import graph.

There are few potential ways to support this:
1. Have `A.register(B)` invalidate all modules that have been already type checked and that may use `A` and  `B`. This can be tricky to implement and could slow down things, as we may have to process modules multiple times. Conservatively, `A.register(B)` would invalidate every module that uses `A` or `B` or a subclass of `B`.
2. Like (1), but also cache all register calls from the previous run and somehow use these to tweak the order of processing modules. The goal is to avoid reprocessing modules multiple times. It's unclear if this idea would work.
3. Make `A.register(B)` visible/active only in modules that (recursively) import the module that has this declaration. Mypy would have to keep track of these register calls defined in each module and which of them are active in any given module. This would deviate from Python semantics and could be a little confusing. However, there would no 'spooky action at a distance'.
4. Have a new pass that only finds `A.register(B)` declarations in the entire program. This would get run before type checking. We can probably make this reasonably fast in incremental mode by being clever with caching etc. [This would probably result in mypy having to keep the entire program in memory when doing a non-incremental run, which would be unfortunate. Currently we could strip most nodes in ASTs of already processed modules to lower RAM use.]
5. Only respect those register calls that mypy has seen at any given stage of processing. This would approximate how these work at runtime, but since mypy can process modules in a different order from Python, this wouldn't always work as expected and could be confusing. This would be easy to implement, at least. We'd probably want to process all register calls inside a module before processing the rest of the module, though, as otherwise the order of definitions within a module would affect type checking results, which would be confusing.

**Comment by JukkaL** [comment]
It seems likely that mypy will support structural subtyping before it supports registering ABCs. Structural subtyping/protocols can be used for a similar effect, but that should be easier to implement. See https://github.com/python/typing/issues/11 for more information.

**Comment by ddfisher** [comment]
Alternatively, instead of doing something very complicated to make sure `A.register(B)` works everywhere, just make `A.register(B)` a type error in modules that don't define either A or B, because you really shouldn't be doing that.  (And it's pretty common to disallow such things, see for example Rust's rules against orphaned impls.)  Solution 3 also seems potentially reasonable.  Spooky action at a distance rules out all the other potential solutions IMO.

**Comment by pkch** [comment]
I just ran into a use case where support for `A.register(B)` would be the only way to make things work:

```
from typing import *
class A('Iterable[A]'):
    _a: 'List[A]'

    def __iter__(self) -> 'Iterator[A]':
        return iter(self._a)

a = A()
reveal_type(a)  # Revealed type is 'test.A'
for x in a:
    reveal_type(x)  # Revealed type is 'test.A*'
```

`mypy` is happy with this - surprisingly given that we don't support recursive types yet (#731). Of course, python runtime isn't too happy that we inherit from a string literal.

The basic problem is that we hacked the class inheritance syntax to do typing-related stuff. But this means we are limited in what we can do - and in particular, we can't use forward declarations there (as well as other constructs that would mess up inheritance; at present, I'm not aware of any specific situations where this comes up).

Since forwarding syntax is already set in stone, the only solution I see is to make this work:

```
class A:
    _a: 'List[A]'
    def __iter__(self) -> 'Iterator[A]':
        return iter(self._a)
Iterable[A].register(A)```
```
It looks a little ugly, but at least it's logical.

True, this particular use case will disappear once Protocols (#3132) is merged. But it won't fully solve the problem because sometimes the user might want to rely on nominal rather than structural typing.

Incidentally, this situation already appears in `builtins.pyi`:

```
class str(Sequence[str]):
    # ...
```

It works there because it's a stub, not the original source code, so python runtime doesn't need to look at it.

(Btw, we should probably make mypy catch inheritance from a string literal before the runtime complains.)

**Comment by ilevkivskyi** [comment]
@pkch 
```python
class A(Iterable['A']):
    ...
```
works both in mypy and at runtime.

Btw, what are the cases where you think people will want nominal checks instead of protocols?

**Comment by pkch** [comment]
Duh 😆 

Nominal checks: if the method naming is too generic (`validate`, `save`, etc.) so the user has (or is afraid to have) a bunch of unrelated classes that would look compatible based on purely structural checks.

**Comment by mensinda** [comment]
Any news/updates on this? I would have really liked support for `.register()` in https://github.com/mesonbuild/meson/pull/8885.

**Comment by ktbarrett** [comment]
@ilevkivskyi Nominal checking is necessary for supporting the numeric tower. It is totally reasonable and common to have subtypes that only exist to bound the set of values of the supertype. Some examples of that:

```python
class natural(int):
    ...

class bit(logic):
    ...

class identifier(str):
    ...
```

**Comment by ktbarrett** [comment]
@ilevkivskyi 
> Btw, what are the cases where you think people will want nominal checks instead of protocols?

Numeric tower inter-operating with the builtin types requires nominal typing and requires this functionality. People wanting to create nominal subtypes that do not want inherit their parent's implementation also requires this functionality. I have a need for the second reason.

```python
class Logic:
    """4-value logic type as seen in Verilog and VHDL: 0, 1, X, and Z"""

class Bit(Logic):
    """2-value logic type: 0 and 1"""

class LogicArray(Sequence[Logic]):
    ...

class BitArray(Sequence[Bit]):
    ...

# can't inherit since the generic base class needs to be re-parameterized
# don't want to anyway since BitArray is implemented with an integer since that's a whole lot faster than a list of user-defined types
LogicArray.register(BitArray)
```

**Comment by pranavrajpal** [comment]
Most of the use cases I'm seeing here are for cases where people are using `ABCMeta.register` on classes they control instead of trying to register externally-defined classes (usually because they want to register a class with an ABC without implementing everything). It would probably be easier to support that case than handling full support for `ABCMeta.register`.

Special casing uses of `register` as a class decorator would most likely be the easiest to implement since that automatically ensures that a class is registered as soon as it's defined and we already have some code for special casing decorators. We'd then probably:
- Emit an error if the class didn't implement all the attributes of the ABC being registered (but we'd only do this when the class is being registered, not when the class is instantiated)
- Treat the class as if that ABC is a base class, even if that earlier unimplemented attributes check failed

**Comment by NeilGirdhar** [comment]
> Special casing uses of register as a class decorator would most likely be the easiest to implement

Isn't that the least important case since anyone who wants to support this can simply make the ABC a base class?

The biggest use case, I think, for this feature would be the numeric tower working as intended.

**Comment by pranavrajpal** [comment]
> Isn't that the least important case since anyone who wants to support this can simply make the ABC a base class?

The important difference between using `register` and making the ABC a base class is when an error about the class not properly implementing all abstract methods gets emitted. If the ABC is a base class, then that happens when the class is actually instantiated, instead of when you initially subclass the ABC, because mypy has to assume that you might be creating an ABC which would implement those methods later in a subclass. However, if you use `register`, we can emit the error at that `register` call itself because the registered class doesn't become an ABC, which means that there's no real reason to register a class with an ABC and then later implement any unimplemented attributes instead of just registering the final class when it has all the methods implemented.

> The biggest use case, I think, for this feature would be the numeric tower working as intended.

My understanding of #3186 is that the builtin number types don't properly implement the numeric tower types that they claim to implement (like in https://github.com/python/typeshed/issues/3195#issuecomment-540946000), and runtime works around that by registering the classes with the correct ABCs.

In that context, having those builtin number types directly subclass the relevant ABCs isn't a realistic option because that would most likely cause errors about unimplemented methods everywhere that those number types are used, even though essentially all of that code is safe.

Using `register` in the way that I mentioned, however, would be an option because we'd emit an error only once when the class was first registered with the ABC. That means that typeshed could register each builtin number type with the correct ABC by adding `register` as a decorator, `type: ignore` the error generated, and anyone who wants to use those builtin number types  use them without getting an error.

**Comment by NeilGirdhar** [comment]
> the builtin number types don't properly implement the numeric tower types that they claim to implement (like in https://github.com/python/typeshed/issues/3195#issuecomment-540946000),

Is this still true?  `float.__ceil__` exists in Python 3.10.  Are there any other examples?

**Comment by NeilGirdhar** [comment]
@ktbarrett 

> Of course the real issue here is that inheritance is a sucky mechanism since it violates separation of concerns. I'd much rather say "Sfixed is a subtype of LogicArray without implementation inheritance".

I think you can just add in the appropriate interface class, and your example will work:
```python
class Logic:
    """4-value logic type as seen in Verilog and VHDL: 0, 1, X, and Z"""

class Bit(Logic):
    """2-value logic type: 0 and 1"""

class LogicArrayInterface(Sequence[Logic]):  ...  # define any members you want to share.

class LogicArray(LogicArrayInterface):
    """Array of Logics that constructs literals or subtypes to enforce the invariant"""

class Sfixed(LogicArrayInterface, Sequence[Bit]):
    """Two's complement arbitrarily-bounded fixed point number as seen in VHDL"""
```
No need for registration.

**Comment by ktbarrett** [comment]
@NeilGirdhar That's not equivalent. Sfixed *is* a subtype of LogicArray by its nature. Not only is it intuitively not correct, it will create more work.

For example, lets say you have a bitwise OR operation that should work on all `Sequence[Logic]` types. The example above shows only two classes, in reality there are 10. What's the resulting type of a bitwise OR operation between a LogicArray and an Sfixed? Intuitively, it would be the less derived type, essentially doing implicit upcasting on the more derived type, so it would return a LogicArray. How would you write that?

```python
class LogicArray(LogicArrayInterface):
    """Array of Logics that constructs literals or subtypes to enforce the invariant"""

    # I don't know the concrete type of the other argument so I don't know the correct type to return.
    def __or__(self, other: LogicArrayInterface) -> ???:
        ...

    # This would work, but it requires that Sfixed be a subtype of LogicArray, that is no longer the case.
    # I can get this to work at runtime using `ABC.register`, but it isn't supported statically.
    def __or__(self: Self, other: Self) -> Self:
        ...
    def __ror__(self: Self, other: Self) -> Self:
        ...

    # Oh dear... this does work though.
    @overload
    def __or__(self, other: LogicArray) -> LogicArray: ...
    @overload
    def __or__(self, other: X01ZArray) -> LogicArray: ...
    @overload
    def __or__(self, other: BitArray) -> LogicArray: ...
    @overload
    def __or__(self, other: Signed) -> LogicArray: ...
    @overload
    def __or__(self, other: Unsigned) -> LogicArray: ...
    @overload
    def __or__(self, other: Sfixed) -> LogicArray: ...
    @overload
    def __or__(self, other: Ufixed) -> LogicArray: ...
    @overload
    def __or__(self, other: Float) -> LogicArray: ...
    # "let's add another type!"
    # *audible groans*
    # *someone posts something about the expression problem*
```

Also, there's a problem in Sfixed with the inheritance hierarchy you have set, `Sequence[Logic]` comes before `Sequence[Bit]`, so those methods will not be correct. If you swap the order of the classes in the inheritance list, you get an ordering issue. You've found another issue with the type system I have also seen playing around with this stuff, where you can't "re-bind" generic classes in subtypes.

**Comment by NeilGirdhar** [comment]
> For example, lets say you have a bitwise OR operation that should work on all Sequence[Logic] types. The example above shows only two classes, in reality there are 10. What's the resulting type of a bitwise OR operation between a LogicArray and an Sfixed? Intuitively, it would be the less derived type, essentially doing implicit upcasting on the more derived type, so it would return a LogicArray. How would you write that?

I see now.  Your problem is analogous to numpy arrays, which are generic on the "dtype".  Essentially, `LogicArrayInterface` is `numpy.ndarray`.

As for your particular problem, can you not simply do:
```python
class LogicArrayInterface:
    @overload
    def __or__(self, other: Self) -> Self:
        ...
    @overload
    def __or__(self, other: LogicArrayInterface) -> LogicArrayInterface:
        ...
```
If you need more clever type promotion, then I'd look at how `numpy` did it.

**Comment by hauntsaninja** [comment]
Summarizing the state of the world + conversation here:

Jukka's comment here still very much applies: https://github.com/python/mypy/issues/2922#issuecomment-283038234. There is not a great way to support `register`

If you control your type, you can simply inherit from the ABC you want:
```
class MyInt(numbers.Integral): ...
```

If you can't actually inherit at runtime, you can lie to the type checker:
```
if TYPE_CHECKING:
    MyIntBase = numbers.Integral
else:
    MyIntBase = object

class MyInt(MyIntBase): ...
```

If you want mypy to check that your MyInt is actually a conformant implementation, you can simply instantiate it:
```
class MyInt(numbers.Integral): ...

if TYPE_CHECKING:
    _ = MyInt()
```

**Comment by ktbarrett** [comment]
@NeilGirdhar 

> LogicArray is like numpy.ndarray and should be generic

That will also not work, unfortunately. Subtypes have more operations on them. For example, `Sfixed + int` is defined, `LogicArray + int` is not. My situation is exactly the same as Python's numeric tower and number types (in fact, if I cared I could integrate my types into the numeric tower). If there was a workaround beyond hacking it into type system like all type checkers currently do with `int` and `float`, or adding support for `ABC.register` no one has figured that out yet.

If one of Jukka's solutions is on the table, it's fairly obvious that 3 is the safest choice and would not have an appreciable effect on performance. The type checker being more strict than runtime already has precedent in typeshed.

**Comment by NeilGirdhar** [comment]
> If one of Jukka's solutions is on the table, it's fairly obvious that 3 is the safest choice and would not have an appreciable effect on performance. 

Totally agree with shooting for number 3.

**Comment by tungol** [comment]
> Special casing uses of register as a class decorator would most likely be the easiest to implement

> Isn't that the least important case since anyone who wants to support this can simply make the ABC a base class?

Support for register as a class decorator could be useful in typeshed, where it would be a bit similar to the `StubOnlyBase` decorator proposed in https://github.com/python/typing/issues/956

**Comment by randolf-scholz** [comment]
Due to divergences in type-checkers, especially overload handling, it is sometimes practically impossible to write a `Protocol` with the desired functionality and assignability that makes multiple target type-checkers happy. In such cases, `register` support would be very handy.

**Comment by DrizztSama** [comment]
Hi everybody. I'm new to mypy (an typing in Python in general) so please excuse any awkwardness in my message.
It seems to me that this issue is concerned with the problem of ABC.register not functioning with mypy when registering two user defined classes but what if we want to register a `builtins` class?

I have in my code objects that represent functions over some known type of stuff and in my code I wanted `float` to represent constant functions over that stuff. So I made an ABC for all things that I am prepared to use as a function over this stuff and registered `float` and it worked as intended. However now that I have decided to type my codes and use mypy I have the same problem as discussed in this issue but worse because I cannot make `builtins.float` derive explicitly from my ABC.

Is there any hope for this situation to be handled one day? If not (or in the meantime), what would you recommend? 

Thanks for reading me

**Comment by alexshpilkin** [comment]
@DrizztSama Assuming your ABC is not exposed externally (other than in type hints) and consequently that the set of types registered with it but not inheriting from it is known in advance, you could use a workaround where you create an alias for a union that includes the ABC and any registered types and use that in your type hints instead.

**Comment by dchevell** [comment]
Would it be acceptable (or useful) to introduce a stricter form of `register` that had to be declared at the same time as the class definition, and not try to solve the problem of searching for calls to `A.register(B)` or rerunning previous checks? A decorator that had to be used on an ABC would effectively mean the class definition and virtual subclass registration were tied together, and could be type checked more sanely.

A simple / toy example:


```python
from abc import ABC, _abc_register 
from typing import Callable


def register_subclass[T: type[ABC]](subclass: type[object]) -> Callable[[T], T]:
    def _register(cls: T) -> T:
        _abc_register(cls, subclass)
        return cls
    return _register


@register_subclass(int)
class MyClass(ABC):
    pass
```

That seems less onerous for typecheckers to special case, and would still preserve the same runtime behaviour in the simple / straight forward use case.

**Comment by ktbarrett** [comment]
@dchevell `ABC.register` can already be used in this way. I agree this would be much easier to support initially.
